internal/app/osd/internal/reg/upgrade.go

```diff
@@ -0,0 +1,120 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package reg
+
+import (
+	"context"
+	"os"
+	"path"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+	"github.com/talos-systems/talos/internal/app/osd/proto"
+	"github.com/talos-systems/talos/internal/pkg/blockdevice/probe"
+	"github.com/talos-systems/talos/internal/pkg/constants"
+	"github.com/talos-systems/talos/internal/pkg/install"
+	"github.com/talos-systems/talos/internal/pkg/kubernetes"
+	"github.com/talos-systems/talos/internal/pkg/mount"
+	"github.com/talos-systems/talos/internal/pkg/version"
+	"golang.org/x/sys/unix"
+)
+
+// Upgrade initiates a Talos upgrade ... and implements the proto.OSDServer
+// interface
+func (r *Registrator) Upgrade(ctx context.Context, in *proto.UpgradeRequest) (data *proto.UpgradeReply, err error) {
+	// Compare current running talos version against specified version
+	newVersion := semver.New(in.Version)
+	currentVersion := semver.New(version.Tag)
+
+	// -1 less than | 0 equal | 1 greater than
+	if newVersion.Compare(*currentVersion) != 1 {
+		return
+	}
+
+	if err = upgradeRoot(in); err != nil {
+		return
+	}
+	if err = upgradeBoot(in); err != nil {
+		return
+	}
+
+	// cordon/drain
+	var kubeHelper *kubernetes.Helper
+	if kubeHelper, err = kubernetes.NewHelper(); err != nil {
+		return
+	}
+
+	var hostname string
+	if hostname, err = os.Hostname(); err != nil {
+		return
+	}
+
+	if err = kubeHelper.CordonAndDrain(hostname); err != nil {
+		return
+	}
+
+	// reboot
+	var initClient *InitServiceClient
+	if initClient, err = NewInitServiceClient(); err != nil {
```

Comment:
I think we already have this client on the `Registrator` struct. `r.Reboot()` should work.